frontend/src/main/java/org/abs_models/frontend/typechecker/locationtypes/LocationTypeExtension.java

```diff
@@ -1,157 +1,139 @@
-/** 
- * Copyright (c) 2009-2011, The HATS Consortium. All rights reserved. 
+/**
```

Comment:
Can/should this old implementation be removed?

Reply:
The entire file? Possibly. It is currently used. After inference, the entire model is checked again by the LocationTypeExtension, which ensures that all inferred types are correct. It also is used when inference is disabled.

Maybe the inference system can cover all checks, but I think it is better to check twice and to leave inference optional.

Reply:
Ok, got it.  That said, if location type inference mutates the AST, maybe it should not be a type checker extension at all?  We have the horrible `Main.analyzeFlattenAndRewriteModel` method that mangles the AST in all kinds of ways; any mutation of the AST should be in there.

(The `Main.analyzeFlattenAndRewriteModel` method is the product of unifying and refactoring all kinds of code; I wouldn't be surprised if another refactoring pass would deconstruct and move it again, but I haven't been in that part of the code for a while.)

Reply:
It doesn't really mutate the AST. It simply adds metadata to the `Type` object.